crawl-ref/source/player-stats.cc

```diff
@@ -97,28 +97,22 @@ int innate_stat(stat_type s)
 
 static void _handle_stat_change(stat_type stat);
 
-/**
- * Handle manual, permanent character stat increases. (Usually from every third
- * XL.
- *
- * @return Whether the stat was actually increased (HUPs can interrupt this).
+/*
+ * Opens a prompt to choose a stat, and returns the stat_type chosen.
+ * If increase_attribute is true, ensure player chooses a stat. If they
+ * do not, restart the prompt.  If choice fails or is cancelled, return
+ * STAT_RANDOM to indicate failure. (hackish, needs better solution?)
  */
-bool attribute_increase()
+stat_type choose_stat(string title, string message, string prompt, bool increase_attribute)
 {
-    // Gnolls don't get stat gains
-    if (you.species == SP_GNOLL)
-        return true;
-
-    const string stat_gain_message = make_stringf("Your experience leads to a%s "
-                                                  "increase in your attributes!",
-                                                  you.species == SP_DEMIGOD ?
-                                                  " dramatic" : "n");
-    crawl_state.stat_gain_prompt = true;
 #ifdef TOUCH_UI
-    learned_something_new(HINT_CHOOSE_STAT);
-    Popup *pop = new Popup("Increase Attributes");
+    if (increase_attribute)
+    {
+        learned_something_new(HINT_CHOOSE_STAT);
+    }
+    Popup *pop = new Popup(title);
```

Comment:
I still don't see a `delete` of this `new`ed Popup or any of the `MenuEntry`s. Is that handled automatically somehow by the touchui code? That would surprise me.

Reply:
If it isn't automatically deallocated by touchui or leaving the function, then the stated problem also exists in the master branch: attribute_increase() does not appear to delete the newed Popup or MenuEntrys after the player makes a choice.  I'll look into it further to see if I can find an answer, and figure out how to add in deletes if they aren't being deallocated already.

Reply:
Added a commit to manually delete the Popup and MenuEntrys when finishing the function.  Leave it out when merging if it turns out they are already being deleted elsewhere.